src/fundamental/memory-util-fundamental.h

```diff
@@ -0,0 +1,42 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+#include <stddef.h>
+
+#ifdef SD_BOOT
+#  include "efi-string.h"
+#else
+#  include <string.h>
+#endif
+
+#include "macro-fundamental.h"
+
+#if defined(HAVE_EXPLICIT_BZERO)
+static inline void *explicit_bzero_safe(void *p, size_t l) {
+        if (p && l > 0)
+                explicit_bzero(p, l);
+
+        return p;
+}
+#else
+static inline void *explicit_bzero_safe(void *p, size_t l) {
+        if (p && l > 0) {
+                memset(p, 0, l);
+                __asm__ __volatile__("" : : "r"(p) : "memory");
```

Comment:
hmm, is this really sufficient? don't we need that volatile memset pointer trick after all?

Reply:
If you must define `explicit_bzero` yourself, I strongly recommend use of this construct instead of any nonsense with `asm`:

```
static inline void explicit_bzero(void *p, size_t l) {
    extern void memory_optimization_fence(void *, size_t);
    if (l > 0)
        memset(p, 0, l);
    memory_optimization_fence(p, l);
}

/* in a separate object file and excluded from LTO */
void memory_optimization_fence(void *unused1, size_t unused2) {}
```

Reply:
Won't that generate an additional `call`? Seems suboptimal.

Having an asm block mark memory as clobbered should have a the same documented effect, right?

Reply:
You shouldn't be trying to micro-optimize this.  The cost of the `memset` is going to dominate no matter what.

The definition of `asm ("" ::: "memory")` is that it _writes_ arbitrary memory (which is strictly worse for optimization than calling a function and providing it with a pointer to a specific block of memory) but there is enough wiggle room in the spec that I don't trust it to also _read_ arbitrary memory, and a read is what's needed here.  I think this is why the godbolt test didn't work.